python/preprocessor.py

```diff
@@ -0,0 +1,349 @@
+remember to update this class if that assumption is changed
+the preprocessor config should be written in the format of
+{
+    name of preprocessor function : [list of observables to apply],
+    other preprocessing tasks
+}
+the preprocessor functions will be run in the same order from top to bottom
+"""
+import json
+import numpy as np
+from collections import OrderedDict
+import gc
+
+import logging
+logger = logging.getLogger('Preprocessor')
+logger.setLevel(logging.DEBUG)
+
+# keys in config
+FEATURE = "feature"
+WEIGHT = "weight"
+
+# observable name indicating all observables will be preprocessed
+all_observable_key = "all_observables"
+
+# preprocessor class
+
+class Preprocessor():
+    def __init__(self, observable_dict, prep_config_path) -> None:
+        """
+        arguments
+        ---------
+        observables: list of str, name of the observables used in training
+        observable_dict: dictionary loaded from observable config
+        prep_config_path: str, path to the preprocessor config file, usually located in configs/preprocessor
+        """
+
+        logger.info("Initializing Preprocessor")
+
+        # map from string options to functions
+
+        self.feature_preprocessing_function_map = {
+            "angle_to_sin_cos": self._angle_to_sin_cos,
+            "angle_to_sin": self._angle_to_sin,
+            "angle_to_cos": self._angle_to_cos
+        }
+
+        weight_preprocessing_function_map = {
+
+        }
+
+        # convert observable dict to a variable dict, mapping branch name to observable name
+        self.observable_name_dict = {}
+        for ob_name in observable_dict:
+            self.observable_name_dict[observable_dict[ob_name]["branch_det"]] = ob_name
+            self.observable_name_dict[observable_dict[ob_name]["branch_mc"]] = ob_name
+
+        # read in config, fill those without assigned preprocessor to an empty list
+        with open(prep_config_path, "r") as config_file:
+            self.config = json.load(config_file, object_pairs_hook=OrderedDict)
+        
+        # save the input on whether the utility functions should be used for outside callers, defaults to False
+        if "preprocess_utility" in self.config[FEATURE]:
+            self.utility = self.config[FEATURE]["preprocess_utility"]
+            # should be removed to not confuse the feature_preprocessing_function_map
+            del self.config[FEATURE]["preprocess_utility"]
+        else:
+            self.utility = True
+
+        self.weight_preprocessing_functions = [weight_preprocessing_function_map[name] for name in self.config[WEIGHT]]
+
+        logger.debug("Initializing Preprocessor: Done")
+
+    # preprocessor functions defined here
+    # they should be of the form
+    # arguments: feature_array, mask, observables, other keyword arguments
+    # returns: preprocessed array, modified observable list (in this order)
+
+    def _angle_to_sin_cos(self, feature_array, mask, observables, **args):
```

Comment:
Does this function, as well as a few others below, need to be a method of the `Preprocessor`? `self` is not really used here.

Reply:
No. We can definitely move them outside of the class.